### ISSUE TITLE
Module graph & pipelines

### ISSUE THREAD

**Issue by muraliran** [NetSys/bess#689]
Following the discussions in #679 I started to look into implementing multiple pipeline. I have a preliminary code base which adds a collection and refactored code to default to default_pipeline for backward compatibility. That was the easy part. In order to implement real multi-pipeline functionality, I wanted to check with you folks what was the original intention around pipelines and flows and graph.

The reason I was looking into this was when I define the reverse path or an egress pipeline wrt ports, it was conflicting with forward flow for attribute reads adding it all in one pipeline. So I was thinking subgraphs can be associated with different pipeline so the attribute scope can be well defined. There are two graphs I see - module graph and the gates path (vector or a graph of gates hook if you will). Pipeline can be a subgraph or how we define paths using gatehook. I see a note to use gatehook as std::map in which case we can associate one or more gate paths as pipeline (similar to module_graph_??). But logically pipeline must cater to multiple flows and I think the gates hook is more like a flow definition. So not sure how to model multiple pipelines. Please share your ideas or provide some feedback on original ideas. Or if my understanding is incorrect, please correct it.

**Comment by muraliran** [comment]
my changes are [here](https://github.com/muraliran/bess/tree/master/core)
See
        modified:   core/bessctl.cc
        modified:   core/hooks/pcapng.cc
        modified:   core/metadata.cc
        modified:   core/metadata.h
        modified:   core/module.cc
        modified:   core/module.h

Also changes set_metadata module to use pipeline

**Comment by melvinw** [comment]
I've only given this a glance so far, but one thing that jumps out at me is the impact this will likely have on performance (even for users who elect not to use the addition).

In particular, [this change to Module::attrs_](https://github.com/muraliran/bess/commit/66d978df8505bd766a03a801dc6d6df003b45399#diff-d805a8bb77d599a990bc63ef59ff85fbR528) worries me. This means finding a metadata attribute within a module now requires two (perhaps only one if you write your modules in a clever way) hash table lookups (one of which is keyed on a string!). That's a considerable amount of extra overhead on top of to the current array of attributes! Perhaps that is worthwhile for some use cases, but it's not clear enough to me exactly what this change buys you, or if it's worthwhile enough, to justify the extra cycles.

**Comment by sangjinhan** [comment]
FYI, the initial intention of "pipeline" was to provide multiple controllers with an isolated dataplane slice. Currently there is no isolation mechanism in the datapath, effectively assuming only one controller is using the datapath (`default_pipeline`).

**Comment by muraliran** [comment]
Melwin, this is not the final code, it is to try the concept of pipeline used & number of attributes is pretty small so not a major concern. The offsets are computed only once on restart. At runtime attributes are looked for one pipeline only..

But it falls apart because pipeline was intended for something else. It is not clear from current code base how pipeline could be used for dataplane isolation because other than attributes it isn't associated with anything.

Anyways I am going to give this some thought if scoping can be done differently for attributes. I would like to have flows defined based on attributes so will be using it a lot.

**Comment by muraliran** [comment]
for the records, I did perf test with and without the pipeline changes I made. The numbers are within the margin of errors, meaning no perf impact. Though I wouldn't have a map with string key in my final code :) 
At this time I have deferred that effort to resolve other stuff, so closed this for now. I get around commenting out attribute descoping lines of code suggested in #679
